pkg/strace/epsocket.go

```diff
@@ -38,71 +39,70 @@ type FullAddress struct {
 	Port uint16
 }
 
+// String implements String
+func (a *FullAddress) String() string {
+	if a == nil {
+		return ":"
+	}
+	return fmt.Sprintf("%s:%#x", []byte(a.Addr), a.Port)
+}
+
 // GetAddress reads an sockaddr struct from the given address and converts it
 // to the FullAddress format. It supports AF_UNIX, AF_INET and AF_INET6
 // addresses.
-func GetAddress(t Task, addr []byte) (FullAddress, error) {
-	r := bytes.NewBuffer(addr[:2])
+func GetAddress(addr []byte) (*FullAddress, error) {
+	r := bytes.NewBuffer(addr)
+
 	var fam uint16
 	if err := binary.Read(r, ubinary.NativeEndian, &fam); err != nil {
-		return FullAddress{}, unix.EFAULT
+		return nil, err
 	}
 
 	// Get the rest of the fields based on the address family.
 	switch fam {
 	case unix.AF_UNIX:
-		path := addr[2:]
+		path := r.Bytes()
+
 		if len(path) > unix.PathMax {
-			return FullAddress{}, unix.EINVAL
+			return nil, unix.ENAMETOOLONG
 		}
+
 		// Drop the terminating NUL (if one exists) and everything after
 		// it for filesystem (non-abstract) addresses.
-		if len(path) > 0 && path[0] != 0 {
-			if n := bytes.IndexByte(path[1:], 0); n >= 0 {
-				path = path[:n+1]
-			}
+		if n := bytes.IndexByte(path, 0); n > 0 {
```

Comment:
If AF_UNIX path doesn't have NUL terminated string it's valid or not? Because from comment it's kinda ok to send it without null terminator `(if one exists)`, but code will return an error in this case.

I'm not sure what is a right answer there, but I would add test case for it and clarify the comment:
```go
{name: "unix no null", addr: []byte{unix.AF_UNIX, 0, 'h', 'i'}, f: &FullAddress{Addr: "hi"}, err: unix.EINVAL}},
```

Also it possible to remove else clause:
```go
		n := bytes.IndexByte(path, 0)
		if n <= 0 {
			return nil, unix.EINVAL
		}

		return &FullAddress{
			Addr: Address(path[:n]),
		}, nil
```

Reply:
looks like it should null terminated `the pathname in sun_path should be null-terminated.`, but it's probably no always the case... So, probably we can check for -1 and return an error?

Reply:
Good questions. This is part of an strace package so we need to manage invalid input -- this is monitoring arbitrary data from user code. The line on 72 does two things:
1. make sure there is a 0 (n will be -1 if not and else is taken)
2. ensure that the 0 is at offset > 0 (paths starting with 0 are not allowed)

in most cases, n will be > 0, meaning a path containing  null that is at least one byte long. 
else return an error. 

On rereading this, I think it's right; am I missing something? 

however, I did add the test you recommended for path not ending in 0. Thanks!

Reply:
Ah got it, thanks for clarification.